beacon-chain/blockchain/head.go

```diff
@@ -438,32 +436,39 @@ func (s *Service) getOrphanedBlocks(ctx context.Context, head [32]byte, base [32
 	return blocks, nil
 }
 
-// Get the most recent common ancestor that both branches are base off.
-func (s *Service) getCommonBase(ctx context.Context, head1 [32]byte, head2 [32]byte) ([32]byte, error) {
-	head1Blk, err := s.cfg.BeaconDB.Block(ctx, head1)
+// Get the latest common ancestor root that both branches are base off.
+func (s *Service) commonAncestorRoot(ctx context.Context, root1 [32]byte, root2 [32]byte) ([32]byte, error) {
+	blk1, err := s.cfg.BeaconDB.Block(ctx, root1)
 	if err != nil {
 		return [32]byte{}, err
 	}
-	head2Blk, err := s.cfg.BeaconDB.Block(ctx, head2)
+	blk2, err := s.cfg.BeaconDB.Block(ctx, root2)
 	if err != nil {
 		return [32]byte{}, err
 	}
+	if blk1 == nil || blk2 == nil {
+		return [32]byte{}, errors.New("nil blocks")
+	}
+
 	// Keep walking back both of the branches until both heads are the same
-	for head1 != head2 {
-		if head1Blk.Block().Slot() >= head2Blk.Block().Slot() {
-			head1 = bytesutil.ToBytes32(head1Blk.Block().ParentRoot())
-			head1Blk, err = s.cfg.BeaconDB.Block(ctx, head1)
+	for root1 != root2 {
+		if ctx.Err() != nil {
```

Comment:
This is necessary because this for loop could go for an infinite amount of time and the requester may have given up already. `ctx.Err()` will return non-nil error when the context timeout has been exceeded or the requester has cancelled the call. Imagine an API user asked for something but when they didn't get a response in 10 seconds then they closed the call. This routine would continue wasting compute cycles to return a response to no one. 

In general, any non-trivial work in a unbounded loop should check for context error so it can exit early.